tests/functional/s3/test_sync_command.py

```diff
@@ -164,3 +165,17 @@ def test_sync_with_delete_on_downloads(self):
         self.assertEqual(self.operations_called[0][0].name, 'ListObjects')
 
         self.assertFalse(os.path.exists(full_path))
+
+    def test_sync_skips_over_files_deleted_between_listing_and_transfer_attempt(self):
+        full_path = self.files.create_file('foo.txt', 'mycontent')
+        cmdline = '%s %s s3://bucket/' % (
+            self.prefix, self.files.rootdir)
+
+        # FileGenerator.list_files should skip over files that cause an OSError to be raised
+        # because they are missing when we try to get their stats.
+        with patch('awscli.customizations.s3.filegenerator.get_file_stat', side_effect=OSError()):
+            self.run_cmd(cmdline, expected_rc=0)
```

Comment:
So unfortunately even if we gets this patch merged in, there is no guarantee that your sync script will always return 0. If you are rotating logs around while syncing them, there is a chance it gets listed out and queued but gets deleted before the upload begins causing that individual file to change.

However it does resolve the reference issue where removing an individual file mid-sync cause the entire sync command to fail. It is much better improvement to have it fail for that individual file instead of the entire sync.

Does that makes sense? Have you tried your patch on your current setup and do you get zero every time now?